### ISSUE TITLE
ARPing timeouts in bridged network setup

### ISSUE THREAD

**Issue by geostarling** [tailhook/lithos#17]
With bridged_network setup, while checking for an veth interface availability on [setup_network](https://github.com/tailhook/lithos/blob/master/src/bin/lithos_knot/setup_network.rs#L199) the arping call fails with a timeout as shown by following output of `lithos_tree` process:

``` text
--- 10.0.0.1 statistics ---
1 packets transmitted, 0 packets received, 100% unanswered (0 extra)

Fatal error: arping failed: exited with code 1
ARPING 10.0.0.1
Timeout

--- 10.0.0.1 statistics ---
1 packets transmitted, 0 packets received, 100% unanswered (0 extra)

Fatal error: arping failed: exited with code 1
ARPING 10.0.0.1
Timeout

```

TCP dump from host (bridge interface) shows no ARP response being sent:

``` text
tcpdump -i br0 -en "icmp or arp"
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on br0, link-type EN10MB (Ethernet), capture size 262144 bytes
15:15:05.564340 4a:41:e6:41:9f:a9 > ff:ff:ff:ff:ff:ff, ethertype ARP (0x0806), length 58: Request who-has 10.0.0.1 (ff:ff:ff:ff:ff:ff) tell 10.0.0.1, length 44
15:15:06.900662 82:84:a5:0c:48:b1 > ff:ff:ff:ff:ff:ff, ethertype ARP (0x0806), length 58: Request who-has 10.0.0.1 (ff:ff:ff:ff:ff:ff) tell 10.0.0.1, length 44
15:15:08.253697 ce:91:ed:2d:a7:19 > ff:ff:ff:ff:ff:ff, ethertype ARP (0x0806), length 58: Request who-has 10.0.0.1 (ff:ff:ff:ff:ff:ff) tell 10.0.0.1, length 44
15:15:09.583953 ba:3f:bc:a6:ac:e8 > ff:ff:ff:ff:ff:ff, ethertype ARP (0x0806), length 58: Request who-has 10.0.0.1 (ff:ff:ff:ff:ff:ff) tell 10.0.0.1, length 44
15:15:10.930642 3a:f1:e9:d9:3f:23 > ff:ff:ff:ff:ff:ff, ethertype ARP (0x0806), length 58: Request who-has 10.0.0.1 (ff:ff:ff:ff:ff:ff) tell 10.0.0.1, length 44
^C
5 packets captured
5 packets received by filter
0 packets dropped by kernel
```

bridge info on host:
``` text
br0       Link encap:Ethernet  HWaddr 00:00:00:00:00:00
          inet addr:10.0.0.200  Bcast:0.0.0.0  Mask:255.255.255.0
          inet6 addr: fe80::1440:b0ff:fec7:5926/64 Scope:Link
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:1047 errors:0 dropped:0 overruns:0 frame:0
          TX packets:727 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:62207 (60.7 KiB)  TX bytes:59939 (58.5 KiB)

```

veth info in the lithos container:

``` text
nsenter -n -p --target 1647 /bin/sh
ifconfig -a
li-ca02f9-0001 Link encap:Ethernet  HWaddr 5A:23:45:F5:F9:B4
          inet addr:10.0.0.1  Bcast:0.0.0.0  Mask:255.255.255.0
          inet6 addr: fe80::5823:45ff:fef5:f9b4/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:12 errors:0 dropped:0 overruns:0 frame:0
          TX packets:10 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:956 (956.0 B)  TX bytes:760 (760.0 B)

lo        Link encap:Local Loopback
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:16 errors:0 dropped:0 overruns:0 frame:0
          TX packets:16 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1
          RX bytes:1376 (1.3 KiB)  TX bytes:1376 (1.3 KiB)
```

I have a trouble understanding what is this check trying to accomplish because:
1. the address (here 10.0.0.1) used in the arping call is that of the veth interface located inside of the container (in child namespace)
2. arping is invoked from the same namespace as the interface is located (ie. child namespace)

That means that veth device with 10.0.0.1 address assigned (in the container) effectively sends ARP broadcast asking "Tell me who has 10.0.0.1" to the bridge (located on host) that the other part of veth is attached and (as expected) receives no answer (the only one that can answer this, is the local interface itself that sends the link broadcast, please note that in this case `arping` beahaves differently than eg. `ping` which _allows_ pinging local interface).

I suspect that the arping call should be performed on the *host (parent) namespace* (possibly with specific bridge interface selected, what if different hosts with the 10.0.0.1 address are reachable from multiple *host*'s NICs?) instead of child namespace, so that it will send ARP from bridge to veth and not the other way around it happens now.

Please see attached patch for better understanding of my proposal:

``` diff
diff --git a/src/bin/lithos_knot/setup_network.rs b/src/bin/lithos_knot/setup_network.rs
index f8df4a6..db9227d 100644
--- a/src/bin/lithos_knot/setup_network.rs
+++ b/src/bin/lithos_knot/setup_network.rs
@@ -193,23 +193,27 @@ fn _setup_bridged(sandbox: &SandboxConfig, _child: &ChildInstance, ip: IpAddr)
             Ok(s) if s.success() => {}
             Ok(s) => bail!("ip route failed: {}", s),
             Err(e) => bail!("ip route failed: {}", e),
         }
     }

+    setns(parent_ns.as_raw_fd(), CloneFlags::CLONE_NEWNET)?;
+
     let mut cmd = unshare::Command::new("/usr/bin/arping");
     cmd.arg("-U");
     cmd.arg("-c1");
     cmd.arg(&format!("{}", ip));
     debug!("Running {}", cmd.display(&Style::short()));
     match cmd.status() {
         Ok(s) if s.success() => {}
         Ok(s) => bail!("arping failed: {}", s),
         Err(e) => bail!("arping failed: {}", e),
     }

+    setns(my_ns.as_raw_fd(), CloneFlags::CLONE_NEWNET)?;
+
     Ok(())
 }

 fn _setup_isolated(_sandbox: &SandboxConfig, _child: &ChildInstance)
     -> Result<(), Error>
 {

```

I have observed this error on Alpine v3.7 *after* PR #15 applied.

Thank you for any ideas or opinions on this.

BTW many thanks to you and other contributors for an awesome containerization stuff (not just lithos). It saves us a lot of time and sanity by not having to deal with Docker :).

**Comment by tailhook** [comment]
Hi,

Thank you for good reports.

Firstly, it think it's just difference how different arping implementations behave. Here is the output of mine:
```
$ arping -U -c1 10.39.0.5
ARPING 10.39.0.5 from 10.39.0.5 li-1683f8-0005
Sent 1 probes (1 broadcast(s))
Received 0 response(s)
```
As you can see, no responses are expected.. I think alpine's implementation doesn't catch that, and timeout is expected. Can you try adding `-w0` or `-w1` ?

I'm not a network expert, but here is how I've been explained it works:

1. Container sends "unsolicited" (-U) arp reponse having a mac address for 10.0.0.1
2. It's just like replying on arp packet for `10.0.0.1`, except no request has been done yet.
3. This packet is broadcasted inside the local network
4. Routers are happy to receive unsolicited arp packets and update a MAC address for the IP `10.0.0.1` (but maybe not all of them...), they are also probably update a port where IP is located

Note: this command is only needed because IP address could be recenly used on a different physical server and is there to update arp cache.

You might try lithos without this ping (v0.15.5) if it works, try quickly reuse IP on different host, and figure out how to make it work (usually connecting to the same IP on a new host takes ~ 5 seconds, if arp cache is not fresh). Maybe we can turn it off entirely or add `-w0` or maybe I'm misusing arping. 

> I suspect that the arping call should be performed on the host (parent) namespace (possibly with specific bridge interface selected, what if different hosts with the 10.0.0.1 address are reachable from multiple host's NICs?) instead of child namespace, so that it will send ARP from bridge to veth and not the other way around it happens now.

I'm not an expert. But hopefully the way it should work is packet having a MAC address and IP of the container and it should reach router. (I believe kernel is smart enough to update routing table internally). So sending a packet from host network namespace will neither container container's MAC nor reach router, if I understand correctly.

Well, I'm still not sure why tcpdump doesn't show the packet for you, so I could be wrong on everything in this comment, but in our network this specific implementation worked and helped to make IPs available immediately.

**Comment by tailhook** [comment]
BTW, are you trying it on a shared hosting/co-location or cloud service or VM? It likely that some platforms ban some arp requests or bridging without some kind of packet encapsulation.

**Comment by geostarling** [comment]
> Firstly, it think it's just difference how different arping implementations behave. Here is the output of mine:

Aha! That seems to be the root of the problem. Alpine's `arping` returns exit code of 1 when the unsolicited ARP request timeouts. The iputils' arping exits with 0 in these cases.

I will try to ask Alpine's `arping` author if there is any chance in harmonizing this behavior with the other implementation.

In the meanwhile, I will probably resort to some form of workaround (BTW is there a chance in making arping notification in lithos non-essential, ie. not leading to failure when the `arping` call fails?).

> Routers are happy to receive unsolicited arp packets and update a MAC address for the IP 10.0.0.1 (but maybe not all of them...), they are also probably update a port where IP is located

Thanks, I have totally missed the main point of using arping in this case (ie. arp cache update via unsolicited request).

Also I have made some further investigation and it seems that some distros (in my case both Fedora and Alpine) by default are not set to update their ARP caches based on unsolicited ARP requests.

As it turns out all you have to do is enable [arp_accept](https://www.kernel.org/doc/Documentation/networking/ip-sysctl.txt) kernel parameter eg. like this:

```
sysctl -w net.ipv4.conf.all.arp_accept=1
```

On my distros this option is disabled by default (arp_accept=0) which causes arp cache to not being updated after lithos process startup:


1. Running `arp` on the host shows only address of default gw of host's NIC (eth0):
```
$ arp
Address                  HWtype  HWaddress           Flags Mask            Iface
gw-10-38-4-1  ether   1c:e6:c7:53:33:c0   C                     eth0
```

2. Switching to network namespace on container and sending ARP request:
```
$ lithos_ps
2476 tree /etc/lithos/master.yaml [1/10] 39.2MiB
└─grafana [1/10] 39.2MiB
└─2512 grafana/grafana.0 9s [1/10] 39.2MiB
$ nsenter -n -p --target 2512 /bin/sh

$ arping -c1 -U 10.0.0.1
ARPING 10.0.0.1
Timeout

--- 10.0.0.1 statistics ---
1 packets transmitted, 0 packets received, 100% unanswered (0 extra)
```

3. Back on the host running `arp` still shows us only the address of the gateway
```
$ arp
Address                  HWtype  HWaddress           Flags Mask            Iface
gw-10-38-4-1  ether   1c:e6:c7:53:33:c0   C                     eth0
```

If I try to `ping` router's address from the container's network namespace, the arp cache on the host gets updated according to expectations:

```
$ ping -c1 10.0.0.200
PING 10.0.0.200 (10.0.0.200): 56 data bytes
64 bytes from 10.0.0.200: seq=0 ttl=64 time=0.254 ms
--- 10.0.0.200 ping statistics ---
1 packets transmitted, 1 packets received, 0% packet loss
round-trip min/avg/max = 0.254/0.254/0.254 ms
```
On host:

```
$ arp
Address                  HWtype  HWaddress           Flags Mask            Iface
10.0.0.1                 ether   02:32:4a:7e:ce:b9   C                     br0
gw-10-38-4-1.in.fit.cvu  ether   1c:e6:c7:53:33:c0   C                     eth0
```


> BTW, are you trying it on a shared hosting/co-location or cloud service or VM? It likely that some platforms ban some arp requests or bridging without some kind of packet encapsulation.

For the record, I'm using KVM-based VM as a host on our local infrastructure.

**Comment by tailhook** [comment]
> BTW is there a chance in making arping notification in lithos non-essential, ie. not leading to failure when the arping call fails?

I'm fine with adding an option to skip arping, but ignoring error code sounds too fragile. 

----

Just to be clear:

> Running arp on the host shows only address of default gw of host's NIC (eth0):

"host" here is a VM running lithos or a host system running outside of KVM?

----

Overall, do you think normal ping to default gateway would work better than arp ping in this case?

**Comment by geostarling** [comment]
> I'm fine with adding an option to skip arping, but ignoring error code sounds too fragile.

Agreed. Also I've just noticed that by making arping to send ARP `responses` by running it with a "-P" option, the timeout does not happen and arp cache on host gets updated properly (after setting aforementioned kernel param).

```
arping -P -U -c1 10.0.0.1
ARPING 10.0.0.1
58 bytes from 5e:63:00:b7:cb:4f (10.0.0.1): index=0 time=10.042 usec

--- 10.0.0.1 statistics ---
1 packets transmitted, 1 packets received,   0% unanswered (0 extra)
```

```
tcpdump -i br0 -en "icmp or arp"
listening on br0, link-type EN10MB (Ethernet), capture size 262144 bytes
15:24:14.681783 5e:63:00:b7:cb:4f > ff:ff:ff:ff:ff:ff, ethertype ARP (0x0806), length 58: Reply 10.0.0.1 is-at 5e:63:00:b7:cb:4f, length 44
```

Unfortunately, it seems that this option is only present in Alpine's arping :(.

> "host" here is a VM running lithos or a host system running outside of KVM?

The host here means KVM VM in which lithos runs (and which provides bridge that lithos container's veths are being attached to).

> Overall, do you think normal ping to default gateway would work better than arpping in this case?

Might be (I'm definitely not an expert). Or better yet, just send a regular (non-unsolicited) ARP requests to default GW address (from container namespace) since in some networking setups ICMP traffic might be blocked.

**Comment by tailhook** [comment]
Well, I have another two observations:
1. `net.ipv4.conf.all.arp_accept=0` in my deployments too
2. Bare `arp` command on my hosts contain no arp entries inside the bridged network (it might be because we don't actually connect to the containers from host system)

>  just send a regular (non-unsolicited) ARP requests to default GW address

This is asking for MAC address of a gateway. I.e. the function of this request is different. It might happen that arp entry for a container is updated as a side-effect, similarly to how ping works, but I'm not sure. Does it update? Even if it is. Is it the most obvious/right way to do it?

-----

Maybe we need to put a command-line for arp cache update into a configuration?

**Comment by geostarling** [comment]
> Bare arp command on my hosts contain no arp entries inside the bridged network (it might be because we don't actually connect to the containers from host system)

That was the case for my deployment as well before I enabled the `arp_accept` parameter. Now I see more ARP entries for other hosts on network (please note that I arp_accept parameter I've mentioned earlier enables this for on _all_ iterfaces)

> Does it update?

Yes it does, see:

1. running from a conainer, 10.0.0.200 is default gw address on the bridge iface
``` text
$ arping -U -c1 10.0.0.200
ARPING 10.0.0.200
42 bytes from ca:51:8b:63:8d:c6 (10.0.0.200): index=0 time=12.852 usec

--- 10.0.0.200 statistics ---
1 packets transmitted, 1 packets received,   0% unanswered (0 extra)
rtt min/avg/max/std-dev = 0.013/0.013/0.013/0.000 ms

```

now on host, we can see address of container being (re)learned from ARP request

``` text
$ arp -i br0
Address                  HWtype  HWaddress           Flags Mask            Iface
10.0.0.1                 ether   36:31:24:95:11:23   C                     br0
```

and on container, we can see address of the gateway updated as well

``` text
$ arp -i li-ca02f9-0001
Address                  HWtype  HWaddress           Flags Mask            Iface
10.0.0.200               ether   ca:51:8b:63:8d:c6   C                     li-ca02f9-0001
10.38.3.17                       (incomplete)                              li-ca02f9-0001
10.38.3.16                       (incomplete)                              li-ca02f9-0001
```


> Is it the most obvious/right way to do it?

I'm not sure there is "the right way" (ie. most generic/robust way) to update ARP cache on listening hosts.
Linux restricts accepting unsolicited ARP requests for a reason (see [ARP spoofing](https://en.wikipedia.org/wiki/ARP_spoofing)). OTOH it is probably less concern in this specific case  (bridged network) as they are used in containers, so in my setup I'm perfectly fine in enablin the arp_accept parameter for the bridge interface.

> Maybe we need to put a command-line for arp cache update into a configuration?

It might be sufficient to allow users to specify paths to custom binary representing for arping (default being iputils' implementation of arping). This way, anyone using non standard impl. of arping in their containers (or just wanting to implement a custom way to perform arp cache update) can use their custom script that will act as an adapter (or in place of) for their implementation of arping.

An ability to specify path to binaries would be useful for other networking utilities as well. Eg. lithos expects `ip` and `brctl` utilities to be located in `/bin/ip` and `/sbin/brctl` respectively. But in Alpine, they are located in `/sbin/ip` ([see](https://pkgs.alpinelinux.org/contents?file=ip&path=&name=&branch=edge)) and `/usr/sbin/brctl` ([see](https://pkgs.alpinelinux.org/contents?file=brctl&path=&name=&branch=edge))

One possible complication might arise, when paths to binaries on container differ from those on the host (so we would need two sets of configurations, one for host and other container specific).

**Comment by tailhook** [comment]
>> Bare arp command on my hosts contain no arp entries inside the bridged network (it might be because we don't actually connect to the containers from host system)
> 
> That was the case for my deployment as well before I enabled the arp_accept parameter. Now I see more ARP entries for other hosts on network (please note that I arp_accept parameter I've mentioned earlier enables this for on all iterfaces)

My point was that we don't need it enabled. It works anyway (unless I miss some edge case).

> This way, anyone using non standard impl. of arping in their containers (or just wanting to implement a custom way to perform arp cache update) can use their custom script

Well, if custom script should behave like arping but isn't arping (i.e. an ICMP ping) it looks weird. It's better to just allow configuring command-line.

By the way, I consider supporting CNI plugins at some point. Do you have some opinion on it?

> An ability to specify path to binaries would be useful for other networking utilities as well

Sounds good. The only downside is if we ever get this functionality built in into lithos. I.e. I believe some `libbridge` crate will appear on crates.io and we use that instead of invoking a binary. This might introduce a non-backwards compatible change.

> One possible complication might arise, when paths to binaries on container differ from those on the host (so we would need two sets of configurations, one for host and other container specific).

We don't use these tools in container's filesystem. We setup both host and container's network namespace using host system's binaries, so we don't need two configs.

----
With all that said, I think simplest solution looks like this:

```
bridged-network:
  bridge: br0
  network: 10.0.0.0/24
  default_gateway: 10.0.0.1
  after-setup-script: [/usr/bin/arping, -U, -c1, '@{container_ip}']
```
Sounds good?

**Comment by geostarling** [comment]
> > > Bare arp command on my hosts contain no arp entries inside the bridged network (it might be because we don't actually connect to the containers from host system)
> >
> > That was the case for my deployment as well before I enabled the arp_accept parameter. Now I see more ARP entries for other hosts on network (please note that I arp_accept parameter I've mentioned earlier enables this for on all iterfaces)
> >
> My point was that we don't need it enabled. It works anyway (unless I miss some edge case).

Out of curiosity: How can you tell that it works when your arp command on host shows no entries? I thought that updating host's ARP cache is the reason for making the arping call from the container.

> > This way, anyone using non standard impl. of arping in their containers (or just wanting to implement a custom way to perform arp cache update) can use their custom script
>
> Well, if custom script should behave like arping but isn't arping (i.e. an ICMP ping) it looks weird. It's better to just allow configuring command-line.

 Agreed.

> By the way, I consider supporting CNI plugins at some point. Do you have some opinion on it?

I have no former experience with the CNI but after a quick glance it looks probably like an overkill for my currently intended deployment (cannot tell at the moment as I'm still evaluating lithos).

It does not look bad as an alternative for a simpler bridged network setups, though I wouldn't want it as a mandatory dependency of lithos.

> > An ability to specify path to binaries would be useful for other networking utilities as well
>
> Sounds good. The only downside is if we ever get this functionality built in into lithos. I.e. I believe some libbridge crate will appear on crates.io and we use that instead of invoking a binary. This might introduce a non-backwards compatible change.

Yes, I recognize that bridged networking in lithos is still in an early state. This specific problem is not an issue for me. I can always symlink the utillities to their expected destinations.


> With all that said, I think simplest solution looks like this:
> Sounds good?

Great, I think that would do the trick!

Also, thanks for quick and helpful replies.

**Comment by tailhook** [comment]
> Out of curiosity: How can you tell that it works when your arp command on host shows no entries? I thought that updating host's ARP cache is the reason for making the arping call from the container.

Connecting to the IP right after container restart becomes possible. If we don't do `arping -U` any packets sent to that IP are getting lost for some time. This takes ~ 5 seconds or so, I don't remember exact timings.

> Great, I think that would do the trick!

Current status: pushed to master, waiting for test package being built to deploy on the test cluster.

**Comment by geostarling** [comment]
Amazing! Thank you very much for your support and lightning fast resolution. I will try it out as soon as possible.

**Comment by tailhook** [comment]
Okay. Works fine for me. Please let me know your results (So that we do less releases with partial functionality).

Also you might want to take a look [here](https://lithos.readthedocs.io/en/latest/sandbox_config.html#bopt-after-setup-command) and 4e47752 to make sure the intent of the option and examples are described clearly.

**Comment by geostarling** [comment]
Works for me too on Alpine with following setup:
`after-setup-command: [/usr/sbin/arping, -U, -P, -c1, '@{container_ip}']`
without timeout and with proper ARP cache update.

As for the third example from the documentation:

> `[/usr/bin/ping, -c1, '10.0.0.1']` – doing same as (2) but using ICMP instead of ARP directly

This would make more sense to me, if lithos allowed variable replacement for `@{default_gateway}` as well. What do you think?

**Comment by tailhook** [comment]
> Works for me too on Alpine with following setup:
after-setup-command: [/usr/sbin/arping, -U, -P, -c1, '@{container_ip}']
without timeout and with proper ARP cache update.

Great! Will tweak docs.

> This would make more sense to me, if lithos allowed variable replacement for @{default_gateway} as well. What do you think?

You can use yaml anchor:
```
bridged-network:
  bridge: br0
  network: 10.0.0.0/24
  default_gateway: &default_gateway 10.0.0.1
  after-setup-command: [/usr/bin/ping, -c1, *default_gateway]
```
Do you see any downsides of this?

**Comment by geostarling** [comment]
> Do you see any downsides of this?

No, this is fine for me.

**Comment by tailhook** [comment]
Released v0.18.0
